hal/src/common/dmac/channel.rs

```diff
@@ -0,0 +1,313 @@
+    ///
+    /// A `Channel` with a `Busy` status.
+    pub(crate) fn start(
+        mut self,
+        trig_src: TriggerSource,
+        trig_act: TriggerAction,
+        dmac: &mut DMAC,
+    ) -> Channel<Busy, ID> {
+        // Set the channel ID. We assume the CHID register doesn't change
+        // for the duration of this function.
+        self.with_chid(dmac, |d| {
+            // Configure the trigger source and trigger action
+            // SAFETY: This is actually safe because we are writing the correct enum value
+            // (imported from the PAC) into the register
+
+            #[cfg(any(feature = "samd11", feature = "samd21"))]
+            let trigger_channel = &d.chctrlb;
+
+            #[cfg(any(
+                feature = "samd51",
+                feature = "same51",
+                feature = "same53",
+                feature = "same54"
+            ))]
+            let trigger_channel = &d.chctrla;
+
+            unsafe {
+                trigger_channel.modify(|_, w| {
+                    w.trigsrc()
+                        .bits(trig_src as u8)
+                        .trigact()
+                        .bits(trig_act as u8)
+                });
+            }
+
+            // Start channel
+            d.chctrla.modify(|_, w| w.enable().set_bit());
+        });
+
+        // If trigger source is DISABLE, manually trigger transfer
+        if trig_src == TriggerSource::DISABLE {
+            self._trigger_private(dmac);
+        }
+
+        Channel { _status: Busy }
+    }
+}
+
+/// These methods may only be used on a `Busy` DMA channel
+impl<const ID: u8> Channel<Busy, ID> {
+    /// Issue a software trigger to the channel
+    pub(crate) fn software_trigger(&mut self, dmac: &mut DMAC) {
+        self._trigger_private(dmac);
+    }
+
+    /// Stop transfer on channel whether or not the transfer has completed
+    ///
+    /// # Return
+    ///
+    /// A `Channel` with a `Ready` status, ready to be reused by a new
+    /// [`DmaTransfer`](super::transfer::DmaTransfer)
+    #[inline(always)]
+    pub(crate) fn stop(mut self, dmac: &mut DMAC) -> Channel<Ready, ID> {
+        self.with_chid(dmac, |d| d.chctrla.modify(|_, w| w.enable().clear_bit()));
+        self.release(dmac)
+    }
+
+    /// Returns whether or not the transfer is complete.
+    ///
+    /// BUSYCH is set when the channel is ACTIVELY transferring;
+    /// PENDCH is set when a trigger request has been received
+    /// but the transfer hasn't been started yet.
+    /// Therefore, when a trigger request is issued, PENDCH will be set first,
+    /// then when the arbiter begins to service the channel, PENDCH is cleared
+    /// and BUSYCH is set. To make sure the transfer is actually complete, the
+    /// channel needs to be both NOT PENDING and NOT BUSY.
+    #[inline(always)]
+    pub(crate) fn xfer_complete(&self, dmac: &mut DMAC) -> bool {
+        dmac.busych.read().bits() & (1 << ID) == 0 && dmac.pendch.read().bits() & (1 << ID) == 0
```

Comment:
Does checking that `BUSYCH` and `PENDCH` are cleared guarantee that the transfer is complete?